scrape/manager.go

```diff
@@ -40,18 +40,23 @@ func NewManager(logger log.Logger, app Appendable) *Manager {
 		scrapeConfigs: make(map[string]*config.ScrapeConfig),
 		scrapePools:   make(map[string]*scrapePool),
 		graceShut:     make(chan struct{}),
+		targetsAll:    make(map[string][]*Target),
 	}
 }
 
 // Manager maintains a set of scrape pools and manages start/stop cycles
 // when receiving new target groups form the discovery manager.
 type Manager struct {
-	logger        log.Logger
-	append        Appendable
-	scrapeConfigs map[string]*config.ScrapeConfig
-	scrapePools   map[string]*scrapePool
-	mtx           sync.RWMutex
-	graceShut     chan struct{}
+	logger         log.Logger
+	append         Appendable
+	scrapeConfigs  map[string]*config.ScrapeConfig
+	scrapePools    map[string]*scrapePool
+	targetsActive  []*Target
+	targetsDropped []*Target
+	targetsAll     map[string][]*Target
+	graceShut      chan struct{}
+	mtxTargets     sync.Mutex // For scrapped targets when access from the web handler.
```

Comment:
I updated the comment to make its purpose more clear.
Don't want to  address the  fields names directly, but rather explain why we need a separate mutex. 

This way we don't need to change the comment  If we add or remove fields.

Reply:
I want you to explicitly list the field names, I'm not going to attempt to review this for correctness until I know what you're trying to do.

Reply:
Many times I find such comment outdated when changing field names around.
Can you explain why is this important?

I find the comment quite to the point.
```
// Mutex for the scrapped targets holders.
// We don't use the  global `mtx` to avoid blocking while reloading the scrape pools.
mtxTargets sync.Mutex
```

Reply:
Otherwise I have to go through the entire file to determine what each mutex is protecting, reverse engineer the intention, verify if that is all correct, and then consider any proposed change. I don't have the time to do all that on every single change, it's much easier when the intention is clearly documented.

Reply:
Following this logic we should add explicit field names for every mutex in every struct.

Anyway , although I disagree , no point in arguing over something like this so I updated the comment.

Reply:
Well, yes. It should be obvious from every struct what mutex applies to what. It's still not clear which fields `mtx` is meant to apply to. I'd suggest grouping fields with their mutex.

Reply:
Please let me know how you want the comments or if possible just change them to your liking.
Otherwise we are wasting time for such unimportant nits and this is dis-motivating.

Reply:
You can do something like the annotations Clang uses for example: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#guarded-by-c-and-pt-guarded-by-c

This is not an unimportant nit. While I appreciate you working on this, the first two attempts both had locking issues. With an even more complicated locking setup in this version I'm not going to even attempt to review for a third time it until the intention with the locks is clearly documented. This is both to make it easier for me as the reviewer, and so that future developers touching this code are far less likely to mess it up.

Reply:
IIUC your idea is something like:
```
scrapeConfigs  map[string]*config.ScrapeConfig // Guarded by mtx
scrapePools    map[string]*scrapePool // Guarded by mtx
targetsActive  []*Target // Guarded by mtxTargets
targetsDropped []*Target  // Guarded by mtxTargets
targetsAll     map[string][]*Target  // Guarded by mtxTargets
```

Reply:
Yes, or anything else that gets the same idea across.

Reply:
updated

while reviewing if you have any other idea how to simplify I don't mind refactoring it again.